### ISSUE TITLE
(pipelines): limit assets concurrency

### ISSUE THREAD

**Issue by rix0rrr** [aws/aws-cdk#17744]
### Description

There are limits to how many CodeBuild projects can run in parallel (it's either 5 or 20), and asset publishing is liable to hit those.

Rather than queueing builds, any builds over the limit are just outright failed.

The CodePipeline generator should take this into account when generating asset publishing stages.

### Use Case

Hitting CodeBuild limit.

### Proposed Solution

Property to limit asset concurrency.

### Acknowledge

- [ ] I may be able to implement this feature request
- [ ] This feature might incur a breaking change

**Comment by gasser-mamdouh** [comment]
I'm really interested to work on this task, can I work on this task?

if I can, for example: if a user has a max of 5 parallel runs, and has 8 assets, I want to make the assets stage have 2 steps the first step have from FileAsset1 to FileAsset5, and the second step have from FileAsset6 to FileAsset8, that's the expected behaviour right?

**Comment by kadrach** [comment]
Does the pipeline builder not already do this?

https://github.com/aws/aws-cdk/blob/588ddf1b509029c70eaf60d0cd852bdc834a3caa/packages/%40aws-cdk/pipelines/lib/codepipeline/codepipeline.ts#L410-L413

The default quota for concurrent CodeBuild jobs is 60 per region, and the maximum number of parallel steps in a CodePipeline stage is 50.

**Comment by volkanunsal** [comment]
Is there a workaround? We have more than 20 lambdas in this service, and I cannot make any headway because of this issue.  Is it possible to adjust this limit? 

@rix0rrr

**Comment by ilgiznurgaliev** [comment]
any updates on this?

**Comment by bchex** [comment]
I am getting an error with our CDK CodePipeline that says  `Error calling startBuild: Cannot have more than 20 builds in queue for the account`. 

There is a service limit for "Concurrently running builds", but it's set to 60 by default. I've already had that limit increased by support twice, and still getting the errors. The [AWS Docs](https://docs.aws.amazon.com/codebuild/latest/userguide/builds-working.html) are relatively clear, but my experience shows that CodePipeline isn't honoring the new limits. I'm guessing that the limits for ComputeType (we're using Large) are usurping the account limits? 

I think this issue will solve our problem but it's not clear. Would really like to see some traction on this.

**Comment by NiallJoeMaher** [comment]
Is there any workaround for this? I've asked for a quota increase but would be nice to have a solution while waiting around if there is one. 
Thanks! 🙏

**Comment by pfried** [comment]
I never actually had this issue despite having a rather large pipeline (large instance) with a lot of lambda assets (split across 2 stages). I came back from vacation and did a deploy and voila: "Error calling startBuild: Cannot have more than 10 builds in queue for the account".

Maybe we never had that many changes (I doubt) or something changed recently

**Comment by DankersW** [comment]
Same error here. 
`Error calling startBuild: Cannot have more than 1 builds in queue for the account (Service: AWSCodeBuild; Status Code: 400; Error Code: AccountLimitExceededException; Request ID: 2cd62e77-6dc1-4c51-9480-9243fba52c91; Proxy: null)`

Nothing changed on our configuration side. Pipelines that have many lambdas do not run any longer.

**Comment by vgjenks** [comment]
Having the same problem. Only allowing 1 build and forcing you to request a limit increase through AWS, is sort of ridiculous. Who can build even a small, mostly pointless pipeline with only 1 concurrent build!? At the very least, we should be able to configure this and accept the billing consequences. It renders CodePiplelines useless to me, at this point.

**Comment by mouhassan** [comment]
The `CodePipeline` construct has a property ([`publishAssetsInParallel`](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.pipelines.CodePipeline.html#publishassetsinparallel)) that should help avoiding this problem.

Setting it to `false` should result in a single CodeBuild project for publishing all assets. This avoids the concurrency problem if the speed of publishing assets is not a concern.

**Comment by byongwu** [comment]
> The `CodePipeline` construct has a property ([`publishAssetsInParallel`](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.pipelines.CodePipeline.html#publishassetsinparallel)) that should help avoiding this problem.
> 
> Setting it to `false` should result in a single CodeBuild project for publishing all assets. This avoids the concurrency problem if the speed of publishing assets is not a concern.

Maybe, it's only me but I am still having failure like this.

```
Error calling startBuild: Cannot have more than 1 builds in queue for the account (Service: AWSCodeBuild; Status Code: 400; Error Code: AccountLimitExceededException; Request ID: cfeca6db-082d-4b33-a4fb-938c72c3d3ff; Proxy: null)
```

<img width="1098" alt="image" src="https://github.com/aws/aws-cdk/assets/69205204/efa52e17-3ad3-4168-ac59-a1b51d28fe05">

Also, I applied the [publishAssetsInParallel](https://docs.aws.amazon.com/cdk/api/v2/docs/aws-cdk-lib.pipelines.CodePipeline.html#publishassetsinparallel) option like this.

```typescript
    const pipeline = new pipelines.CodePipeline(this, "Pipeline", {
      pipelineName: props.codepipelineName,
      synth: steps,
      publishAssetsInParallel: false,
    });
```

**Comment by Gum-Christopher-bah** [comment]
Hoping this comment helps someone else if they find it. I found this thread after trying to look into ways to speed up our pipeline deployments, as cdk was generating hundreds of these assets steps. In total, even with parallel execution, this took a minimum of 30 minutes with our setup. Setting `publishAssetsInParallel: false` along with increasing the codebuild instance size did it in one execution in about 3 minutes.

**Comment by fukunaga-kengo** [comment]
Hi, I am interested in working on this issue. Before starting, I would like to confirm the intended API and behavior.

Would the preferred approach be to add an optional property that sets the maximum number of asset publishing actions that may run concurrently? In particular, could a maintainer clarify:

- whether the limit should apply across both file and Docker assets, or separately to each asset type;
- whether a limit of `1` should serialize file and Docker asset publishing as well;
- how the new setting should interact with the existing `publishAssetsInParallel` property; and
- whether assets beyond the limit should be placed into subsequent CodePipeline run orders, for example 8 assets with a limit of 5 becoming groups of 5 and 3.

I would preserve the current behavior when the new property is not specified. If this is still the desired direction, I would be happy to prepare an implementation.
